java-checks/src/main/java/org/sonar/java/checks/VarCanBeUsedCheck.java

```diff
@@ -0,0 +1,135 @@
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.java.checks;
+
+import java.util.Collections;
+import java.util.List;
+import java.util.Locale;
+
+import org.sonar.check.Rule;
+import org.sonar.java.JavaVersionAwareVisitor;
+import org.sonar.java.model.ExpressionUtils;
+import org.sonar.java.model.JUtils;
+import org.sonar.plugins.java.api.IssuableSubscriptionVisitor;
+import org.sonar.plugins.java.api.JavaFileScannerContext;
+import org.sonar.plugins.java.api.JavaVersion;
+import org.sonar.plugins.java.api.semantic.Type;
+import org.sonar.plugins.java.api.tree.ExpressionTree;
+import org.sonar.plugins.java.api.tree.IdentifierTree;
+import org.sonar.plugins.java.api.tree.LiteralTree;
+import org.sonar.plugins.java.api.tree.MemberSelectExpressionTree;
+import org.sonar.plugins.java.api.tree.MethodInvocationTree;
+import org.sonar.plugins.java.api.tree.SyntaxToken;
+import org.sonar.plugins.java.api.tree.Tree;
+import org.sonar.plugins.java.api.tree.TypeTree;
+import org.sonar.plugins.java.api.tree.VariableTree;
+
+@Rule(key = "S6212")
+public class VarCanBeUsedCheck extends IssuableSubscriptionVisitor implements JavaVersionAwareVisitor {
+
+  public static final String MESSAGE = "Declare this local variable with \"var\" instead.";
+  private int typeAssignmentLine = -1;
+
+  @Override
+  public List<Tree.Kind> nodesToVisit() {
+    return Collections.singletonList(Tree.Kind.VARIABLE);
+  }
+
+  @Override
+  public boolean isCompatibleWithJavaVersion(JavaVersion version) {
+    return version.isJava10Compatible();
+  }
+
+  @Override
+  public void setContext(JavaFileScannerContext context) {
+    typeAssignmentLine = -1;
+    super.setContext(context);
+  }
+
+  @Override
+  public void visitNode(Tree tree) {
+    VariableTree variableTree = (VariableTree) tree;
+    ExpressionTree initializer = variableTree.initializer();
+    TypeTree type = variableTree.type();
+    Type symbolType = type.symbolType();
+    IdentifierTree identifierTree = variableTree.simpleName();
+
+    if (isMultiAssignment(variableTree) ||
+      initializer == null ||
+      type.is(Tree.Kind.VAR_TYPE) ||
+      symbolType.isUnknown() ||
+      !JUtils.isLocalVariable(variableTree.symbol()) ||
+      initializer.is(Tree.Kind.CONDITIONAL_EXPRESSION) ||
+      symbolType.isParameterized()) {
+      return;
+    }
+
+    Type initializerType = initializer.symbolType();
+    if (symbolType.fullyQualifiedName().equals(initializerType.fullyQualifiedName()) &&
+      (isSelfExplained(initializer) ||
+        typeWasMentionedInTheName(identifierTree, symbolType) ||
+        typeWasMentionedInTheInitializer(initializer, symbolType))) {
+      reportIssue(identifierTree, MESSAGE);
+    }
+  }
+
+  private static boolean typeWasMentionedInTheName(IdentifierTree variable, Type type) {
+    return isLogicallyReferable(variable.name(), type.name());
```

Comment:
You call this method many times with the same second argument, and end up doing multiple times the same operation: `type.name().toLowerCase(Locale.ROOT)`. We should compute it only once.